src/lib.rs

```diff
@@ -17,244 +17,147 @@
 )]
 #![forbid(unsafe_code)]
 
-#[cfg(feature = "std")]
-#[macro_use]
-extern crate std;
+use ark_ff::{FpParameters, PrimeField};
+use ark_std::vec::Vec;
+use ark_std::vec;
 
-#[cfg(not(feature = "std"))]
 #[macro_use]
-extern crate alloc as std;
+extern crate derivative;
+
+/// Infrastructure for the constraints counterparts.
+#[cfg(feature = "r1cs")]
+pub mod constraints;
+
+mod absorbable;
+pub use absorbable::*;
 
-use ark_ff::models::{
-    Fp256, Fp256Parameters, Fp320, Fp320Parameters, Fp384, Fp384Parameters, Fp768, Fp768Parameters,
-    Fp832, Fp832Parameters,
-};
-use ark_ff::{to_bytes, PrimeField, ToConstraintField};
-use std::vec::Vec;
+/// The sponge for Poseidon
+///
+/// This implementation of Poseidon is entirely from Fractal's implementation in [COS20][cos]
+/// with small syntax changes.
+///
+/// [cos]: https://eprint.iacr.org/2019/1076
+pub mod poseidon;
+
+/// A sponge that offers backwards compatibility for implementations that do not accept sponge
+/// objects but require domain separation. Operates in the same way as fork.
+pub mod domain_separated;
 
 /// An enum for specifying the output field element size.
-#[derive(Clone)]
+#[derive(Clone, Copy, Eq, PartialEq)]
 pub enum FieldElementSize {
     /// Sample field elements from the entire field.
     Full,
 
-    /// Sample field elements from a subset of the field.
-    Truncated {
-        /// The maximum size of the subset is 2^num_bits.
-        num_bits: usize,
-    },
+    /// Sample field elements from a subset of the field, specified by the maximum number of bits.
+    Truncated(usize),
+}
+
+impl FieldElementSize {
+    pub(crate) fn num_bits<F: PrimeField>(&self) -> usize {
+        if let FieldElementSize::Truncated(num_bits) = self {
+            *num_bits.min(&(F::Params::CAPACITY as usize))
+        } else {
+            F::Params::CAPACITY as usize
+        }
+    }
 }
 
 /// The interface for a cryptographic sponge.
 /// A sponge can `absorb` or take in inputs and later `squeeze` or output bytes or field elements.
 /// The outputs are dependent on previous `absorb` and `squeeze` calls.
-pub trait CryptographicSponge<F: PrimeField> {
+pub trait CryptographicSponge<CF: PrimeField>: Clone {
```

Comment:
remove `<CF: PrimeField>`, and move the type parameter to the methods that need it.

Reply:
I've removed the type parameter by adding an associated type (that is, the native field used by the sponge implementation). @Will-Lin4 can you do a sanity check to confirm if this trick works?

Reply:
I think the idea was that instead of having generic type parameters or associated types, we would move it to the methods themselves, like `squeeze_field_elements<CF: PrimeField>`. This would allow us to combine squeeze field elements and squeeze nonnative field elements into a single function. Does this sound correct @Pratyush?

Reply:
However, in this way, shall we treat native field and non-native field the same (i.e. first convert the field to bytes, and then convert back to field elements)

Reply:
Similar comments for `squeeze_field_elements`, do we treat `non_native` and `native` the same? (I don't think so...)

Reply:
@Pratyush brought up the idea of using type IDs to distinguish between the native or nonnative PrimeFields, but I am not sure what to do about the case with `absorb`. We could do something that always involves checking type IDs but it might make the implementation of different `Absorb`s a bit messier at the expense of making the API cleaner. 

It isn't entirely clear to me how this would be fleshed out.

Reply:
In fact, if we want to implement byte-based crytographic sponge like SHA or blake2s, then `fn absorb(&mut self, input: &impl Absorable<CF>)` does not make sense to me, because those sponges do not seem to have any `CF`. We might need to have a sync to reconsider the API for sponge

Reply:
Also, not sure if I misunderstood, I was a little confused of the type parameter of the trait `Absorbable<CF>` as well. If the user has a struct contains a field element (let's say `Fp256`), then current implementation limits the user to use a sponge based on `Fp256` unless user manually implements `Absorbale<other field>` for that field elements. Also, if we want to implement byte-based sponge in future, then `to_sponge_bytes` is enough, and `CF` type parameter doesn't make sense here.

However, I did not come up with an easy way to drop that `CF` type parameter.  I'm still thinking about a better way to handle that complexity though.

Reply:
I think using the type id looks good, then I believe we can remove the type parameter in `Absorbable`. Here is a proposed new API: 

```rust
pub trait Absorbale{
   fn to_sponge_bytes(&self) -> Vec<u8>;
   fn to_sponge_field_elements<F: PrimeField>(&self) -> Vec<F>;
}
```

In `to_sponge_field_elements`, we can use type ID to check if it is native or not. Is this a good idea? @Will-Lin4

Reply:
I think the original purpose of having `CF` as a part of `Absorbable` was to allow implementations with sponges to not care about whether a byte-based sponge or an algebraic sponge would be used (eg. if a sponge needed to be passed in as a generic type parameter). Any input that is absorbed in the implementation would work regardless because of the two functions.

Yeah, I think that change looks good if existing implementations of `Absorbable` can be updated to work with it.

Reply:
Got it thanks! I will try to update existing implementations next week. After that I will send you an update

Reply:
should be fixed now